server/src/main/java/org/elasticsearch/common/settings/Setting.java

```diff
@@ -366,10 +366,34 @@ public T getDefault(Settings settings) {
     }
 
     /**
-     * Returns <code>true</code> iff this setting is present in the given settings object. Otherwise <code>false</code>
+     * Returns true if and only if this setting is present in the given settings instance. Note that fallback settings are excluded.
+     *
+     * @param settings the settings
+     * @return true if the setting is present in the given settings instance, otherwise false
      */
-    public boolean exists(Settings settings) {
-        return settings.keySet().contains(getKey());
+    public boolean exists(final Settings settings) {
+        return exists(settings, false);
+    }
+
+    /**
+     * Returns true if and only if this setting optionally including fallback settings is present in the given settings instance.
+     *
+     * @param settings the settings
+     * @param fallback whether or not to include fallback settings
+     * @return true if the setting and optionally fallback settings is present in the given settings instance, otherwise false
+     */
+    public boolean exists(final Settings settings, final boolean fallback) {
+        Setting<?> current = this;
+        do {
+            if (settings.keySet().contains(current.getKey())) {
+                return true;
+            }
+            if (fallback == false) {
+                break;
+            }
+            current = current.fallbackSetting;
```

Comment:
Can't you call `current.fallbackSetting.exists(settings, true)` instead of a loop?

Reply:
That is a great suggestion. I pushed 7ae63df9d7aeddec7e4a45d0d561bae6ea882d3a.